### PR TITLE
fix(uploads): recognize valid ATX headings in document outlines

### LINKED ISSUE #5313: [bug] Hashtags and indented code consume uploaded document outline entries
## Summary

**Implementation status:** We have implemented and verified the fix; our PR is #5316. We are continuing the work in that PR. Please coordinate here before starting another implementation to avoid duplicated effort.

The outline extractor treats every stripped line starting with # as a heading. Ordinary hashtags and indented code therefore occupy the limited outline, potentially hiding real document sections from the agent.

## Actual behavior

51 lines containing #tag followed by # Real heading produce 50 false headings and a truncated sentinel. Real heading is missing. Four-space-indented # comments and seven-hash lines are also included; optional closing hashes remain in titles.

## Expected behavior

Recognize root-level ATX headings with 1-6 opening hashes, a space/tab separator or end of line, and at most three leading spaces. Remove valid optional closing hashes. Non-headings must not consume the outline limit.

## Steps to reproduce

On main a3848ef1, write a UTF-8 Markdown file containing '#tag\n' repeated 51 times followed by '# Real heading\n'. Call extract_outline(path) from deerflow.utils.file_outline. Observe that Real heading is absent and the outline is marked truncated. This was reproduced with the unchanged production file in a no-network DockerVerifier container.

## Environment

Linux, Python 3.12; main a3848ef1; synthetic local Markdown, no provider calls.

## Acceptance criteria

- [ ] Hashtags, seven-hash prefixes and indented code are not outline entries.
- [ ] Valid ATX headings retain physical line numbers and clean optional closing markers.
- [ ] Existing fenced-code handling, PDF bold headings and real-heading limit behavior remain covered.

## Additional context

Scope: ATX recognition in the shared document-outline utility and regression coverage. This is independent of the fenced-code fix merged in #5281 and does not implement a full Markdown parser or add Setext/list/blockquote heading support.

Reference: https://spec.commonmark.org/0.31.2/#atx-headings . Open Issues/PRs and matching closed work were checked; no active fix for ATX recognition was found.

AI assistance: Codex investigated source, reproduced the behavior and prepared this report.

### PR DESCRIPTION
**Pull request by tiammomo**
Fixes #5313.

## Why

An uploaded document with 51 `#tag` lines before `# Real section` consumes all outline slots and hides the real section. Four-space/tab-indented comments also become false headings, and optional closing hashes remain in titles.

## What changed

The new long-heading regression exceeds its 10-second process deadline at the previous head. The fixed production extractor passes it; measured median times for 8k/16k/32k/262144 spaces were 0.036/0.044/0.081/0.460 ms in the diagnostic container. The suffix cleanup is linear by construction (bounded str.rstrip passes), without changing valid ATX results.

Match root-level ATX syntax on the original line (1–6 hashes, space/tab separator or end of line, at most three leading spaces), then strip valid closing hashes before existing bold cleanup. Keep physical line numbers, PDF structural headings, fenced-code exclusion and the actual-heading limit. This remains a bounded scanner, not a complete Markdown parser.

## Surface area

- [x] **Default behavior change** — Exclude hashtags and indented code from uploaded document outlines; normalize valid ATX heading markers.
- One production module, regression tests and matching documentation; no dependency changes.

## Bug fix verification

The initial 21 ATX syntax cases in `backend/tests/test_file_outline_atx.py` produced 12 failures and 9 passes before the original syntax fix; all 21 passed afterward. The additional long-heading deadline regression is described above. Covers false-heading overflow, indentation, hash counts, closing markers, escaped markers, Unicode titles and empty headings.

## Validation

- Additional integration diagnostic: the prepared commits `941c82c6` (#5316) and `26904d71` (#5323) merge automatically; all 62 focused outline/ATX/text-budget tests pass on their combined tree. This is separate from each PR’s full verification below.

- Backend: 15457 passed, 167 skipped, 3 deselected, 20 warnings in 691.05s (0:11:31).
- Blocking-I/O: 125 passed, 2 warnings in 6.10s.
- Full relevant lint/format and scope-aware guidance pass. Frontend updates additionally pass TypeScript and a production build.
- Verified commit `941c82c649cf5dde1d7046426dbffc0f7659b97f` on upstream `bc4a33ab`; upstream changes were merged without dropping either side of the documentation.
- Hardened DockerVerifier with locked bootstrap and offline tests. No test retries, exclusions or relaxed assertions added this round.
- The new long-heading regression exceeds its 10-second process deadline at the previous head. The fixed production extractor passes it; measured median times for 8k/16k/32k/262144 spaces were 0.036/0.044/0.081/0.460 ms in the diagnostic container. The suffix cleanup is linear by construction (bounded str.rstrip passes), without changing valid ATX results.

## AI assistance

**Tool(s) used:** Codex.

**How you used it:** Source investigation, offline reproduction, implementation, regression tests and review.

- [x] I've read and understand every line of this change and take responsibility for it — it's not unreviewed AI output.

### CODE REVIEW

**Review by willem-bd — COMMENTED** [review]
Reviewed at cc948ebf9f4e62eff0c96a7282d464ef4d49f79d. Implementation verified by execution: the new suite scores exactly the 12 failed / 9 passed claimed against the base revision, and 47/47 pass at head (21 new + the existing `test_file_outline.py`). I also differential-tested heading detection against `markdown_it` (CommonMark) over ~35 ATX edge cases — no detection mismatches; the only deltas are the pre-existing intentional ones (bold-artifact cleanup, empty-content headings skipped). Two non-blocking notes below: the new indentation guard covers ATX only, and the strictness can silently empty an outline for hand-uploaded Markdown.

File: backend/packages/harness/deerflow/utils/file_outline.py [hunk 1]
```diff
@@ -42,6 +42,11 @@
 # the line is an info string when opening, or whitespace only when closing.
 _CODE_FENCE_RE = re.compile(r"^ {0,3}(`{3,}|~{3,})(.*)$")
 
+# ATX headings require 1-6 hashes and a space/tab separator (or end of line).
+# Match the original indentation so indented code cannot become a heading.
+_ATX_HEADING_RE = re.compile(r"^ {0,3}#{1,6}(?:[ \t]+(.*))?$")
+_ATX_CLOSING_RE = re.compile(r"(?:^|[ \t]+)#+[ \t]*$")
```
Comment (willem-bd) [inline backend/packages/harness/deerflow/utils/file_outline.py:48]: Verified empirically: this guard holds for style 1 only. Styles 2 and 3 still match on `stripped`, so the same indented-code false headings this PR removes for ATX still occupy outline slots via the bold patterns:

```
    # Code comment        -> []                                          (fixed by this PR)
    **ITEM 1. BUSINESS**  -> [{'title': 'ITEM 1. BUSINESS', 'line': 1}]
\t**2** **Not a section** -> [{'title': '2 Not a section', 'line': 1}]
```

If "indented code cannot become a heading" is the intent, the two bold styles need the same raw-line indentation check; otherwise the comment and docstring should scope the claim to ATX headings.

File: backend/packages/harness/deerflow/utils/file_outline.py [hunk 2]
```diff
@@ -118,8 +124,8 @@ def extract_outline(md_path: Path) -> list[dict]:
                     continue
 
                 # Style 1: standard Markdown heading
-                if stripped.startswith("#"):
-                    title = _clean_bold_title(stripped.lstrip("#").strip())
+                if m := _ATX_HEADING_RE.fullmatch(line.rstrip("\r\n")):
+                    title = _clean_bold_title(_ATX_CLOSING_RE.sub("", m.group(1) or "").strip())
```
Comment (willem-bd) [inline backend/packages/harness/deerflow/utils/file_outline.py:128]: Non-blocking: correct per CommonMark, but worth noting the failure mode for hand-uploaded `.md`. Only `.pdf/.ppt/.xls/.doc*` go through markitdown/pymupdf4llm, so a directly uploaded Markdown file that uses `#Heading` (no space), 7+ hashes, or a tab/NBSP separator now yields an empty outline, and `extract_outline_for_file` silently falls back to the first 5 non-empty lines — with nothing in the logs explaining why visible heading structure was dropped. A debug log in that branch (e.g. a count of lines matching `^ {0,3}#{1,6}\S`) would make "why is my outline empty" diagnosable.

File: backend/packages/harness/deerflow/utils/file_outline.py (hunk 1 above)
Comment (tiammomo) [reply backend/packages/harness/deerflow/utils/file_outline.py:48]: The scope here is ATX heading recognition. The indentation comment directly accompanies `_ATX_HEADING_RE`, and the function docstring scopes the three-space rule to “Standard ATX headings”. The two converter-specific bold styles intentionally keep their existing recognition behavior in this PR; a shared indentation rule would change that separate heuristic. Thank you for the independent differential check.

File: backend/packages/harness/deerflow/utils/file_outline.py (hunk 2 above)
Comment (tiammomo) [reply backend/packages/harness/deerflow/utils/file_outline.py:128]: Keeping the existing first-five-nonempty-lines fallback is intentional: text without valid ATX headings still produces a preview. One clarification: ASCII tabs after the hashes are accepted by `[ \t]+` and are covered by the new tests; NBSP is not a CommonMark heading separator. I am keeping diagnostic instrumentation outside this focused parsing fix rather than adding another content scan to uploads.

File: backend/packages/harness/deerflow/utils/file_outline.py (hunk 1 above)
Comment (willem-bd) [inline backend/packages/harness/deerflow/utils/file_outline.py]: [P2] Keep closing-hash cleanup linear on uploaded content

The unanchored `[ \t]+` alternative retries from every whitespace position when no closing hash sequence follows, producing quadratic work. I reproduced this with a one-line Markdown file containing `'# Title' + ' ' * 32000 + 'suffix\n'`: the actual `extract_outline` takes about 2.32 seconds at this head versus 0.00016 seconds on the base. Doubling the whitespace from 8,000 to 16,000 to 32,000 characters grows the cleanup from roughly 0.145 to 0.559 to 2.263 seconds. Heading length is not bounded, and the 50-entry limit does not constrain processing one line, so a small uploaded document can tie up outline extraction for a long time. Please use a linear suffix scan (trim trailing spaces/tabs, inspect the terminal hash run and its separator) or another demonstrably linear implementation, with a regression covering long whitespace runs without a closing marker.

**Review by willem-bd — COMMENTED** [review]
Reviewed at cc948ebf. One P2 performance regression is detailed inline, reproduced with the actual extract_outline function against the base and head revisions. Validation: 170 focused and surrounding tests passed; changed-file Ruff lint/format, diff checks, and scope-aware agent-guidance checks passed.

File: backend/packages/harness/deerflow/utils/file_outline.py (hunk 1 above)
Comment (tiammomo) [reply backend/packages/harness/deerflow/utils/file_outline.py]: Addressed in 941c82c6. Replaced the unanchored closing-marker regex with bounded str.rstrip suffix scans. A production-extractor regression with 262144 spaces exceeded a 10-second deadline at the old head and passes after the fix; fixed diagnostic medians are 0.081 ms at 32000 spaces and 0.460 ms at 262144. Existing ATX/bold/fence behavior remains covered. Fresh offline validation: Backend: 15457 passed, 167 skipped, 3 deselected, 20 warnings in 691.05s (0:11:31); Blocking-I/O: 125 passed, 2 warnings in 6.10s.

**Review by WillemJiang — APPROVED** [review]
